src/charm.py

```diff
@@ -53,92 +55,85 @@ def __init__(self, *args):
             role=self.config["role"],
             installed=False,
             joined=False,
-            leaving=False,
-            join_url="",
             hostnames={},
-            hostname=socket.gethostname(),
         )
 
         if self.config["role"] == "worker":
-            self.framework.observe(self.on.remove, self._on_remove)
-            self.framework.observe(self.on.install, self._on_install)
-            self.framework.observe(self.on.install, self._worker_open_ports)
-            self.framework.observe(self.on.config_changed, self._on_config_changed)
-            self.framework.observe(self.on.update_status, self._on_update_status)
-            self.framework.observe(self.on.microk8s_relation_joined, self._announce_hostname)
-            self.framework.observe(self.on.microk8s_relation_joined, self._retrieve_join_url)
-            self.framework.observe(self.on.microk8s_relation_changed, self._retrieve_join_url)
-            self.framework.observe(self.on.microk8s_relation_broken, self._on_relation_broken)
+            self.framework.observe(self.on.remove, self.on_remove)
+            self.framework.observe(self.on.upgrade_charm, self.on_upgrade)
+            self.framework.observe(self.on.install, self.on_install)
+            self.framework.observe(self.on.update_status, self.update_status)
+            self.framework.observe(self.on.config_changed, self.config_ensure_role)
+            self.framework.observe(self.on.config_changed, self.on_install)
+            self.framework.observe(self.on.config_changed, self.config_containerd_proxy)
+            self.framework.observe(self.on.config_changed, self.config_containerd_registries)
```

Comment:
Do we know the order the on.config_changed hooks will be called? If not it would be better to have a singe hook calling all our functions in a very specific order. I feel this will save us from some future headaches.

Reply:
Yes, they are executed in the order with which they were defined, and this is enforced by the operator framework.